File: src/Functional/Elixir/flyweight.exs
defmodule StylePool do
  def get(pool, key, style) do
    case Map.fetch(pool, key) do
      {:ok, existing} -> {existing, pool}
      :error -> {style, Map.put(pool, key, style)}
    end
  end
end

pool = %{}
{red1, pool} = StylePool.get(pool, {"Inter", 12, "red"}, %{font: "Inter", size: 12, color: "red"})
{red2, pool} = StylePool.get(pool, {"Inter", 12, "red"}, %{font: "Inter", size: 12, color: "red"})
{blue, pool} = StylePool.get(pool, {"Inter", 12, "blue"}, %{font: "Inter", size: 12, color: "blue"})
true = blue.color == "blue"
shared = red1 === red2
IO.puts("styles=#{map_size(pool)};shared=#{shared};text=ABC")

File: src/Functional/Erlang/flyweight.erl
-module(flyweight).
-export([main/0]).

get_style(Pool, Key, Style) ->
    case maps:find(Key, Pool) of
        {ok, Existing} -> {Existing, Pool};
        error -> {Style, maps:put(Key, Style, Pool)}
    end.

main() ->
    Pool0 = #{},
    {Red1, Pool1} = get_style(Pool0, {"Inter", 12, "red"}, #{font => "Inter", size => 12, color => "red"}),
    {Red2, Pool2} = get_style(Pool1, {"Inter", 12, "red"}, #{font => "Inter", size => 12, color => "red"}),
    {Blue, Pool3} = get_style(Pool2, {"Inter", 12, "blue"}, #{font => "Inter", size => 12, color => "blue"}),
    "blue" = maps:get(color, Blue),
    Shared = Red1 =:= Red2,
    io:format("styles=~B;shared=~p;text=ABC~n", [map_size(Pool3), Shared]).
